File: HW1/javaApp/src/main/java/io/rienel/domain/Mouse.java
package io.rienel.domain;

import java.util.Objects;

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
public class Mouse extends Animal {

    private Double tailLength;

    public Mouse(final String name,
                 final Double weight,
                 final Double fullHeight,
                 final Double fullLength,
                 final Double fullWidth,
                 final Double tailLength) {
        super(name, weight, AnimalType.MOUSE, fullHeight, fullLength, fullWidth);
        this.tailLength = tailLength;
    }

    @Override
    public void setName(String name) {
        Objects.requireNonNull(name);
        super.setName(name);
    }

    @Override
    public void setWeight(Double weight) {
        Objects.requireNonNull(weight);
        if (weight < 0.5 || weight > 1.0) {
            throw new IllegalArgumentException("Illegal weight");
        }
        super.setWeight(weight);
    }

    @Override
    public void setFullHeight(Double fullHeight) {
        Objects.requireNonNull(fullHeight);
        if (fullHeight < 2.0 || fullHeight > 10.0) {
            throw new IllegalArgumentException("Illegal fullHeight");
        }
        super.setFullHeight(fullHeight);
    }

    @Override
    public void setFullLength(Double fullLength) {
        Objects.requireNonNull(fullLength);
        if (fullLength < 2.0 || fullLength > 10.0) {
            throw new IllegalArgumentException("Illegal fullLength");
        }
        super.setFullLength(fullLength);
    }

    @Override
    public void setFullWidth(Double fullWidth) {
        Objects.requireNonNull(fullWidth);
        if (fullWidth < 2.0 || fullWidth > 10.0) {
            throw new IllegalArgumentException("Illegal fullWidth");
        }
        super.setFullWidth(fullWidth);
    }

    @Override
    public Boolean eat(Animal food) {
        Objects.requireNonNull(food);

        return tailLength <= 10;
    }

    @Override
    public String speak() {
        return "MOUSE: SQUEAK";
    }
}

File: HW1/kotlinApp/src/main/kotlin/io/rienel/domain/Mouse.kt
package io.rienel.domain

import io.rienel.domain.interfaces.Feedable
import io.rienel.domain.interfaces.Speekable

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
class Mouse(
    name: String,
    weight: Double,
    fullHeight: Double,
    fullLength: Double,
    fullWidth: Double,
    var tailLength: Double
) : Animal(
    name,
    weight,
    AnimalType.MOUSE,
    fullHeight,
    fullLength,
    fullWidth
), Feedable<Animal>, Speekable {

    override var name: String
        get() = super.name
        set(value) {
            assert(value.isNotEmpty())
            super.name = value
        }
    override var weight: Double
        get() = super.weight
        set(value) {
            assert(value in 0.500 .. 1.0)
            super.weight = value
        }

    override var fullHeight: Double
        get() = super.fullHeight
        set(value) {
            assert(fullHeight in 50.0 .. 100.0)
            super.fullHeight = value
        }

    override var fullLength: Double
        get() = super.fullLength
        set(value) {
            assert(fullLength in 50.0 .. 100.0)
            super.fullLength = value
        }
    override var fullWidth: Double
        get() = super.fullWidth
        set(value) {
            assert(fullWidth in 50.0 .. 100.0)
            super.fullWidth = value
        }

    override fun eat(food: Animal): Boolean {
        return tailLength <= 10
    }

    override fun speak(): String {
        return "MOUSE: SQUEAK"
    }
}